File: gtci/12-tree-depth-first-search/cpp/4-sum-of-path-numbers.cpp
#include <cassert>
#include <iostream>

#include "tree_utils.h"

using namespace std;

class Solution {
public:
  int findSumOfPathNumbers(TreeNode *root) {
    if (root == nullptr)
      return 0;
  
    return findSumOfPathNumbersHelper(root, 0);
  }

  int findSumOfPathNumbersHelper(TreeNode* node, int currentNumber) {
    if (node == nullptr)
      return 0;
    
    currentNumber = currentNumber * 10 + node->val;

    if (node->left == nullptr && node->right == nullptr)
      return currentNumber;
    
    return findSumOfPathNumbersHelper(node->left, currentNumber) + findSumOfPathNumbersHelper(node->right, currentNumber);
  }
};

int main() {
    Solution sol;

    TreeNode *root1 = new TreeNode(1);
    root1->left = new TreeNode(7);
    root1->right = new TreeNode(9);
    root1->right->left = new TreeNode(2);
    root1->right->right = new TreeNode(9);
    assert(sol.findSumOfPathNumbers(root1) == 408);
    
    TreeNode *root2 = new TreeNode(1);
    root2->left = new TreeNode(0);
    root2->right = new TreeNode(1);
    root2->left->left = new TreeNode(1);
    root2->right->left = new TreeNode(6);
    root2->right->right = new TreeNode(5);
    assert(sol.findSumOfPathNumbers(root2) == 332);
    
    cout << "All test cases passed." << endl;

    delete root1;
    delete root2;

    return 0;
}

File: gtci/12-tree-depth-first-search/python3/4-sum-of-path-numbers.py
class TreeNode:
    def __init__(self, val, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def findSumOfPathNumbers(self, root):
        if root is None:
            return 0

        return self.findSumOfPathNumbers_helper(root, 0)

    def findSumOfPathNumbers_helper(self, node, current_number):
        if node is None:
            return 0

        current_number = current_number * 10 + node.val

        if node.left is None and node.right is None:
            return current_number

        return self.findSumOfPathNumbers_helper(
            node.left, current_number
        ) + self.findSumOfPathNumbers_helper(node.right, current_number)


if __name__ == "__main__":
    sol = Solution()

    root = TreeNode(1)
    root.left = TreeNode(7)
    root.right = TreeNode(9)
    root.right.left = TreeNode(2)
    root.right.right = TreeNode(9)

    assert sol.findSumOfPathNumbers(root) == 408

    root = TreeNode(1)
    root.left = TreeNode(0)
    root.right = TreeNode(1)
    root.left.left = TreeNode(1)
    root.right.left = TreeNode(6)
    root.right.right = TreeNode(5)

    assert sol.findSumOfPathNumbers(root) == 332

    print("All test cases passed.")
